### DISCUSSION TITLE
Use Docker layer caching with docker-compose build (not just docker)

### DISCUSSION THREAD

**Discussion by valentijnscholten** [community/community#25728] [Actions]
<p>Hi,</p>
<p>I spent some time getting my workflows to work with the docker layer caching as described in the blog by Docker: <a href="https://www.docker.com/blog/docker-github-actions/" class="inline-onebox" rel="noopener nofollow ugc">Docker Github Actions - Docker Blog</a></p>
<p>This works fine for building the images with Docker:</p>
<pre><code class="lang-auto">      - name: Set up Docker Buildx
        id: buildx        
        uses: docker/setup-buildx-action@master

      - name: Cache Docker layers
        uses: actions/cache@v2
        with:
          path: /tmp/.buildx-cache
          key: ${{ runner.os }}-buildx-step1-${{ github.sha }}
          restore-keys: |
            ${{ runner.os }}-buildx-

      - name: Build and push images
        uses: docker/build-push-action@v2
        with:
          push: false
          builder: ${{ steps.buildx.outputs.name }}
          tags: repo/step1:latest
          file: ./Dockerfile.step1
          context: .
          cache-from: type=local,src=/tmp/.buildx-cache
          cache-to: type=local,dest=/tmp/.buildx-cache
</code></pre>
<p>And I have 1 more container build like that (step2).</p>
<p>I have a docker-compose setup to run unit tests &amp; integration tests with a whole bunch of docker-compose specific config around environment variables, volumes, entrypoints, etc.</p>
<p>So I am running these tests using docker-compose. Problem is docker-compose doesn’t use the images built by Docker. I did a lot of searching and googling and tried lots of things, but everything I found doesn’t seem to help to get it to work in GitHub actions.</p>
<p>Latest try (after the docker build steps above):</p>
<pre><code class="lang-auto">      - name: Build the stack
        run: COMPOSE_DOCKER_CLI_BUILD=1 docker-compose build
</code></pre>
<p>Doesn’t help, docker compose starts rebuilding both containers from scratch. Also doing <code>docker-compose up</code> doesn’t help, it starts pulling the images from docker hub.</p>
<p>Anyone knows the solution? Could be that I’m missing something, but most examples of docker layer caching (with github actions) doesn’t seem to cover docker-compose.</p>
<p>Full workflow and runs here: <a href="https://github.com/valentijnscholten/django-DefectDojo/pull/60/checks?check_run_id=1717919129" class="inline-onebox" rel="noopener nofollow ugc">Docker caching by valentijnscholten · Pull Request #60 · valentijnscholten/django-DefectDojo · GitHub</a></p>
<p>Valentijn</p>

**Comment by valentijnscholten** [comment]
<p>Added a basic PoC here where I build with docker-build-push action and then use docker-compose up with --no-build</p>
<p>Locally on my machine this works fine. Docker-compose uses the images build by docker.  On GitHub actions it doesn’t work. the jobs fails as it says the image is not there.</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/valentijnscholten/docker-caching/actions" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:64/64;"><img src="https://avatars2.githubusercontent.com/u/4426050?s=400&amp;amp;v=4" class="thumbnail" width="64" height="64"></div>

<h3><a href="https://github.com/valentijnscholten/docker-caching/actions" target="_blank" rel="noopener nofollow ugc">valentijnscholten/docker-caching</a></h3>

<p><a href="https://github.com/valentijnscholten/docker-caching/actions" target="_blank" rel="noopener nofollow ugc">ns</a></p>

  <p><span class="label1">poc. Contribute to valentijnscholten/docker-caching development by creating an account on GitHub.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by valentijnscholten** [comment, marked as answer]
<p>ok, found the solution. I realized that maybe the docker-build-and-push action didn’t actually do anything with the generated image, i.e. it was not made available to the runner / docker. Turns out this is true and even documented:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/docker/build-push-action#export-image-to-docker" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:400/400;"><img src="https://avatars2.githubusercontent.com/u/5429470?s=400&amp;amp;v=4" class="thumbnail" width="400" height="400"></div>

<h3><a href="https://github.com/docker/build-push-action#export-image-to-docker" target="_blank" rel="noopener nofollow ugc">docker/build-push-action - Export image to Docker</a></h3>


  <p><span class="label1">:octocat: GitHub Action to build and push Docker images with Buildx</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Once you know what you’re looking for, it’s easy…</p>
<p>In case the link dies, the solution is to add <code>load: True</code> to the docker-build-and-push action:</p>
<pre><code class="lang-auto">- name: Build
        uses: docker/build-push-action@v2
        with:
          push: false
          load: true
          tags: valentijn/app:latest
          builder: ${{ steps.buildx.outputs.name }}
          file: ./Dockerfile
          context: .
          cache-from: type=local,src=/tmp/.buildx-cache
          cache-to: type=local,dest=/tmp/.buildx-cache
</code></pre>

**Reply by rknop** [reply]
This doesn't answer the question... it is for an individual Dockerfile, not for all of the images referenced in a docker compose file.

**Comment by mlissner** [comment]
<p>Note that there’s a bug that makes the cache grow and grow if you do it this way. There’s a <a href="https://github.com/docker/build-push-action/blob/9379083e426e2e84abb80c8c091f5cdeb7d3fd7a/docs/advanced/cache.md" rel="noopener nofollow ugc">fix here</a>.</p>

**Comment by splatEric** [comment]
<p>I struggled to get this working - docker-compose would not pick up on the cache content, in spite of following the suggestions around using the <code>load</code> parameter etc.</p>
<p>However, I did find a 3rd party option that might be suitable for anyone having similar problems to me:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/satackey/action-docker-layer-caching" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>
  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:400/400;"><img src="https://avatars.githubusercontent.com/u/21271711?s=400&amp;amp;v=4" class="thumbnail" width="400" height="400"></div>

<h3><a href="https://github.com/satackey/action-docker-layer-caching" target="_blank" rel="noopener nofollow ugc">satackey/action-docker-layer-caching</a></h3>


  <p><span class="label1">🐳  Enable Docker layer caching in GitHub Actions. Contribute to satackey/action-docker-layer-caching development by creating an account on GitHub.</span></p>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>I do recommend using the documented step of pulling images, as the caching was much slower when I omitted it on my first run through. Oddly I found that docker-compose ignored the <code>ignore-pull-failures</code> flag, so I threw in the <code>continue-on-error</code> flag.</p>
<pre><code class="lang-auto">- name: Docker Compose pull
     run: docker-compose pull --ignore-pull-failures
     continue-on-error: true
</code></pre>

**Comment by marcellodesales** [comment]
<p>This is not working … I have build args in docker-compose and it is not providing on the builds… Also, the logs doesn’t show any indication…</p>
<p>I would expect that the library called “<a href="https://docs.docker.com/engine/reference/commandline/buildx_bake/" class="inline-onebox" rel="noopener nofollow ugc">docker buildx bake | Docker Documentation</a>” buildx bake…</p>

**Comment by aretasg-alchemab** [comment]
<p>Awesome! It works. Just don’t keep in mind that the tag in Action workflow and specified image name in the docker-compose.yml have to be the same. Otherwise, it rebuilds the image instead of using cache.</p>
